File: apps/web/convex/mcpCredentials.ts
import { getAuthUserId } from '@convex-dev/auth/server';
import { v } from 'convex/values';

import { internal } from './_generated/api';
import { action, internalQuery } from './_generated/server';

interface CredentialSummary {
  credentialId: string;
  name: string;
  scopes: string[];
  createdAt: string;
  lastUsedAt: string | null;
  expiresAt: string | null;
  revokedAt: string | null;
}

interface CreatedCredential {
  credentialId: string;
  tenantId: string;
  name: string;
  scopes: string[];
  token: string;
  createdAt: string;
  expiresAt?: number;
}

const credentialValidator = v.object({
  credentialId: v.string(),
  name: v.string(),
  scopes: v.array(v.string()),
  createdAt: v.string(),
  lastUsedAt: v.union(v.string(), v.null()),
  expiresAt: v.union(v.string(), v.null()),
  revokedAt: v.union(v.string(), v.null()),
});

export const requireWorkspaceAdmin = internalQuery({
  args: { workspaceId: v.id('workspaces') },
  returns: v.object({ workspaceId: v.string() }),
  handler: async (ctx, { workspaceId }) => {
    const userId = await getAuthUserId(ctx);
    if (!userId) throw new Error('Unauthorized.');

    const member = await ctx.db
      .query('members')
      .withIndex('by_workspace_id_user_id', (query) => query.eq('workspaceId', workspaceId).eq('userId', userId))
      .unique();
    if (!member || member.role !== 'admin') throw new Error('Workspace admin access required.');

    return { workspaceId };
  },
});

async function callCalyx<T>(path: string, body: object): Promise<T> {
  const baseUrl = process.env.CALYX_API_URL;
  const internalKey = process.env.CALYX_INTERNAL_API_KEY;
  if (!baseUrl || !internalKey) throw new Error('Calyx connector management is not configured.');

  const response = await fetch(`${baseUrl.replace(/\/$/, '')}${path}`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      'x-calyx-internal-key': internalKey,
    },
    body: JSON.stringify(body),
  });
  const payload = (await response.json().catch(() => ({}))) as { error?: unknown } & T;
  if (!response.ok) {
    const message = typeof payload.error === 'string' ? payload.error : `Calyx API returned ${response.status}`;
    throw new Error(message);
  }
  return payload;
}

export const list = action({
  args: { workspaceId: v.id('workspaces') },
  returns: v.array(credentialValidator),
  handler: async (ctx, { workspaceId }): Promise<CredentialSummary[]> => {
    const authorization: { workspaceId: string } = await ctx.runQuery(internal.mcpCredentials.requireWorkspaceAdmin, {
      workspaceId,
    });
    const result: { credentials: CredentialSummary[] } = await callCalyx('/v1/mcp/credentials/list', {
      workspaceId: authorization.workspaceId,
    });
    return result.credentials;
  },
});

export const create = action({
  args: {
    workspaceId: v.id('workspaces'),
    name: v.string(),
    expiresInDays: v.optional(v.number()),
  },
  returns: v.object({
    credentialId: v.string(),
    tenantId: v.string(),
    name: v.string(),
    scopes: v.array(v.string()),
    token: v.string(),
    createdAt: v.string(),
    expiresAt: v.optional(v.number()),
  }),
  handler: async (ctx, args): Promise<CreatedCredential> => {
    const authorization: { workspaceId: string } = await ctx.runQuery(internal.mcpCredentials.requireWorkspaceAdmin, {
      workspaceId: args.workspaceId,
    });
    if (args.name.trim().length < 1 || args.name.trim().length > 80) throw new Error('Invalid credential name.');
    if (args.expiresInDays !== undefined && (!Number.isInteger(args.expiresInDays) || args.expiresInDays < 1 || args.expiresInDays > 365)) {
      throw new Error('Expiry must be between 1 and 365 days.');
    }

    const result: { credential: CreatedCredential } = await callCalyx('/v1/mcp/credentials/create', {
      workspaceId: authorization.workspaceId,
      name: args.name.trim(),
      scopes: ['logs:read', 'incidents:read', 'incidents:ask'],
      ...(args.expiresInDays && { expiresInDays: args.expiresInDays }),
    });
    return result.credential;
  },
});

export const revoke = action({
  args: { workspaceId: v.id('workspaces'), credentialId: v.string() },
  returns: v.boolean(),
  handler: async (ctx, args): Promise<boolean> => {
    const authorization: { workspaceId: string } = await ctx.runQuery(internal.mcpCredentials.requireWorkspaceAdmin, {
      workspaceId: args.workspaceId,
    });
    await callCalyx<{ revoked: true }>('/v1/mcp/credentials/revoke', {
      workspaceId: authorization.workspaceId,
      credentialId: args.credentialId,
    });
    return true;
  },
});

File: apps/api/internal/httpapi/mcp_credentials.go
package httpapi

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"
)

func (s *Server) requireWorkspaceAdmin(r *http.Request) (string, bool) {
	workspaceID := chi.URLParam(r, "workspaceID")
	member, err := s.requireMember(r.Context(), workspaceID, userID(r.Context()))
	return workspaceID, err == nil && member.Role == "admin"
}

func (s *Server) proxyMCPCredentials(w http.ResponseWriter, r *http.Request, action string, payload map[string]any) {
	if s.calyxAskURL == "" || s.calyxInternalKey == "" {
		writeErr(w, http.StatusServiceUnavailable, "Calyx connector management is not configured")
		return
	}
	body, _ := json.Marshal(payload)
	req, err := http.NewRequestWithContext(r.Context(), http.MethodPost, s.calyxAskURL+"/v1/mcp/credentials/"+action, bytes.NewReader(body))
	if err != nil {
		writeErr(w, http.StatusInternalServerError, "connector request failed")
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Calyx-Internal-Key", s.calyxInternalKey)
	response, err := s.httpClient.Do(req)
	if err != nil {
		writeErr(w, http.StatusBadGateway, "Calyx connector service unavailable")
		return
	}
	defer response.Body.Close()
	responseBody, err := io.ReadAll(io.LimitReader(response.Body, 2<<20))
	if err != nil {
		writeErr(w, http.StatusBadGateway, "invalid connector response")
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(response.StatusCode)
	_, _ = w.Write(responseBody)
}

func (s *Server) listMCPCredentials(w http.ResponseWriter, r *http.Request) {
	workspaceID, ok := s.requireWorkspaceAdmin(r)
	if !ok {
		writeErr(w, http.StatusForbidden, "workspace admin access required")
		return
	}
	s.proxyMCPCredentials(w, r, "list", map[string]any{"workspaceId": workspaceID})
}

func (s *Server) createMCPCredential(w http.ResponseWriter, r *http.Request) {
	workspaceID, ok := s.requireWorkspaceAdmin(r)
	if !ok {
		writeErr(w, http.StatusForbidden, "workspace admin access required")
		return
	}
	var body struct {
		Name          string `json:"name"`
		ExpiresInDays *int   `json:"expiresInDays"`
	}
	if decodeJSON(r, &body) != nil {
		writeErr(w, http.StatusBadRequest, "invalid json")
		return
	}
	body.Name = strings.TrimSpace(body.Name)
	if body.Name == "" || len(body.Name) > 80 {
		writeErr(w, http.StatusBadRequest, "name must contain 1-80 characters")
		return
	}
	if body.ExpiresInDays != nil && (*body.ExpiresInDays < 1 || *body.ExpiresInDays > 365) {
		writeErr(w, http.StatusBadRequest, "expiry must be between 1 and 365 days")
		return
	}
	payload := map[string]any{
		"workspaceId": workspaceID,
		"name":        body.Name,
		"scopes":      []string{"logs:read", "incidents:read", "incidents:ask"},
	}
	if body.ExpiresInDays != nil {
		payload["expiresInDays"] = *body.ExpiresInDays
	}
	s.proxyMCPCredentials(w, r, "create", payload)
}

func (s *Server) revokeMCPCredential(w http.ResponseWriter, r *http.Request) {
	workspaceID, ok := s.requireWorkspaceAdmin(r)
	if !ok {
		writeErr(w, http.StatusForbidden, "workspace admin access required")
		return
	}
	credentialID := strings.TrimSpace(chi.URLParam(r, "credentialID"))
	if credentialID == "" || len(credentialID) > 100 {
		writeErr(w, http.StatusBadRequest, "invalid credential id")
		return
	}
	s.proxyMCPCredentials(w, r, "revoke", map[string]any{"workspaceId": workspaceID, "credentialId": credentialID})
}
